llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp

```diff
@@ -1428,6 +1428,40 @@ static Instruction *foldBitOrderCrossLogicOp(Value *V,
   return nullptr;
 }
 
+/// Fold an unsigned minimum of trailing or leading zero bits counts:
+///   umin(cttz(CtOp, ZeroUndef), ConstOp) --> cttz(CtOp | (1 << ConstOp))
+///   umin(ctlz(CtOp, ZeroUndef), ConstOp) --> ctlz(CtOp | ((1 << (bitwidth-1))
+///                                              >> ConstOp))
+template <Intrinsic::ID IntrID>
+static Instruction *foldMinimumOverTrailingOrLeadingZeroCount(
+    Instruction *OrigInst, Value *CtOp, Value *ZeroUndef, Constant *ConstOp,
+    const DataLayout &DL, InstCombiner::BuilderTy &Builder) {
+  static_assert(IntrID == Intrinsic::cttz || IntrID == Intrinsic::ctlz,
+                "This helper only supports cttz and ctlz intrinsics");
+
+  auto BitWidth = ConstOp->getType()->getScalarSizeInBits();
+  auto *Ty = ConstOp->getType();
+
+  Constant *NewConst = ConstantFoldSelectInstruction(
+      ConstantFoldCompareInstOperands(CmpInst::ICMP_ULT, ConstOp,
+                                      ConstantInt::get(Ty, BitWidth), DL),
+      ConstantFoldBinaryOpOperands(
+          IntrID == Intrinsic::cttz ? Instruction::Shl : Instruction::LShr,
+          IntrID == Intrinsic::cttz
+              ? ConstantInt::get(Ty, 1)
+              : ConstantInt::get(Ty, APInt::getSignedMinValue(BitWidth)),
+          ConstOp, DL),
+      Constant::getNullValue(Ty));
+
+  Value *NewCtOp =
+      NewConst->isZeroValue() ? CtOp : Builder.CreateOr(CtOp, NewConst);
+
+  return CallInst::Create(Intrinsic::getDeclaration(OrigInst->getModule(),
```

Comment:
Can we not just call ComputeKnownBits on ConstOp and check if Known.getMaxValue().ult(BitWidth)?

Reply:
As far as I understand, `Known.getMaxValue().ult(BitWidth)` would evaluate to `false` for a vector with only some elements greater or equal to the bit width because the different vector values are combined into a single `KnownBits` value. In this case, performing the transformation would make the result potentially poisonous. In contrast, the current implementation handles the case of non-uniform vectors with some elements that are less and some elements that are greater or equal to the bit width by folding an element-wise select operation and therefore also works for these cases.

Reply:
Its fine for this fold to only work for cases with all elements less than bitwidth.

Reply:
An alternative would be to work with @goldsteinn and make use of #85676

Reply:
#85676 has been pushed. Think this could would be simplified by it.

Reply:
The new matchers are exactly what I've needed. I've changed the implementation to make use of `m_CheckedInt`.